rfcs/0000-socket-proxy-kinesis-json.md

```diff
@@ -0,0 +1,168 @@
+
+Kinesis consists of named *streams*, and each stream consists of one or more *shards*. *Records* are read/written to the stream, and distributed between the shards based on a *partition key*. *Producers* write records to the stream, and are read by zero or more *consumers*. When written, each record receives a *sequence number*: this number is always increasing and uniquely identifies the record within the shard. It also orders the messages: messages within a shard have a defined order, but not messages received by two different shards.
+
+In this case, SocketProxy will be the producer. For each message received from the OCS, it will transform that message into a CloudEvent (see [RFC 4](https://github.com/mbta/technology-docs/pull/4)), encode the event as JSON, and put a record in the "ctd-raw-ocs-messages" stream. The "ctd-raw-ocs-messages" stream will be configured with two (2) shards. In order that train movement messages are processed in the correct order, the partition key will be based on the type of OCS message and the line.
+
+The four (4) RTR instances will be consumers of the stream. Instead of expecting messages on TCP sockets, they will read the CloudEvent JSON payload from the stream, parse out the raw message, and send it through the existing pipeline.
+
+A [JSON Schema](https://json-schema.org/) document will be written in the `mbta/schemas` repository, with the name of the CloudEvent message.
+
+# Reference-level explanation
+
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+
+Currently, SocketProxy has a Receiver GenServer responsible for each incoming connection from OCS. Receiver then spawns a Forwarder GenServer to send the data to each of the destinations (currently, the four RTR instances). After this RFC, SocketProxy will have a KinesisReceiver which is responsible for writing records to a configured Kinesis stream:
+
+```
+SOCKET_PROXY_LISTEN_PORT=8000 SOCKET_PROXY_KINESIS_STREAM=ctd-ocs-raw-messages mix run --no-halt
+```
+
+The maximum size of a record (including the partition key) is 1MB, and record storage is charged in units of 25KB. Each shard supports writing 1MB or 1000 records per second For reading, each shard can support 5 read transactions per second, and up to 10,000 records per transaction, up to a total of 2MB per second. One (1) shard should be sufficient, but we will provision two (2) shards. While one shard should be sufficient, having multiple shards ensures that consumers properly handle multiple shards from the start.
```

Comment:
Kinesis does not drop messages: readers can get behind if they don't process messages fast enough. (Thank you for the Splunk query: I was trying to do that myself and couldn't get it to work).

The 5 reads / second might be raise-able: I'll look into it. Even if it's not, however, RTR should be able to fetch all the messages from the last second in one read even during peak times.

It definitely makes sense for the readers to use Braodway/GenState (the `kcl_ex` implementation of the Kinesis Client Library uses Broadway under the hood). But for the writer, it gets messages at the rate that OCS sends them to us: how much is it possible to give them backpressure?

Reply:
Digging in, it looks like TMOVs peak at 50/s. It's the TSCH which peak at 400/s but could also be batched effectively. If the polling is done in batches of 1/s, could code currently in RTR to push GTFS-RT updates every second be removed? Then it would go as fast as the polling rate and we'd limit the delay. I'll also do some local testing to see what the latency looks like with 1/s polling rate: since messages are available immediately after writing, the average should be lower than that.

The 5/s/shard read rate isn't raise-able, but there's another option if it turns out that's not enough. [Enhanced fan-out consumers](https://docs.aws.amazon.com/streams/latest/dev/building-enhanced-consumers-api.html) 1) each one gets a dedicated 2MB/s for each shard 2) messages are pushed via HTTP/2 rather than needing to be polled. The disadvantage is that it costs more: enhanced fan out is charged by both consumer/shard/hour and GB retrieved, and the consumers need to be registered ahead of time.

Reply:
Looks like AWS also recommends using enhanced fan-out here:
![Screen Shot 2021-06-07 at 9 58 06 AM](https://user-images.githubusercontent.com/214921/121029851-f1b26200-c776-11eb-9afb-0c30b609266b.png)

I'll also look into that this week.

Reply:
This does work, but we'll need to write some of our own code to support it.

Reply:
> But for the writer, it gets messages at the rate that OCS sends them to us: how much is it possible to give them backpressure?

Heh, there's a limited capacity to apply backpressure, but at some point it causes things to 💥 in the OCC (ask me how I know). We should definitely architect to drain incoming messages as quickly as possible.

Well, not listening at all won't cause any issues there, but accepting the TCP connection and then not reading from it fast enough can be a problem. So I suppose if necessary we _could_ close the connection if we were falling behind. Then when we reconnect we'd get some of the interim messages, but I don't think OCC will buffer indefinitely.

> If the polling is done in batches of 1/s, could code currently in RTR to push GTFS-RT updates every second be removed?

Yep. It might be a little bit of work to do so. At the moment, receiving messages is async from processing the GTFS-rt file, and I don't know to what extent the test suite (and in particular integration tests) rely on that. I think it would simplify things overall, though.

Regarding shards and fan-outs, I'll leave that up to you. :) But it sounds like it _can_ handle whatever we need.

Reply:
I think we can have the KinesisProxy write data as fast as possible. We know that RTR can currently process the data fast enough, and this shouldn't change that.